pkg/faucet/faucet.go

```diff
@@ -0,0 +1,987 @@
+	if err != nil {
+		return fmt.Errorf("build transaction payload failed, error: %w", err)
+	}
+
+	msg, err := f.createMessage(txPayload, tip...)
+	if err != nil {
+		return fmt.Errorf("build faucet message failed, error: %w", err)
+	}
+
+	messageID, err := f.sendMessageFunc(ctx, msg)
+	if err != nil {
+		return fmt.Errorf("send faucet message failed, error: %w", err)
+	}
+
+	f.Lock()
+	f.lastMessageID = &messageID
+	f.addPendingTransactionWithoutLocking(&pendingTransaction{
+		MessageID:      messageID,
+		QueuedItems:    batchedRequests,
+		ConsumedInputs: consumedInputs,
+		TransactionID:  *transactionID,
+	})
+
+	if remainderIotaGoOutput != nil {
+		remainderIotaGoOutputID := remainderIotaGoOutput.ID()
+		output := &iotago.BasicOutput{
+			Amount: remainderAmount,
+			Conditions: iotago.UnlockConditions{
+				&iotago.AddressUnlockCondition{Address: f.address},
+			},
+		}
+		f.lastRemainderOutput = &UTXOOutput{
+			OutputID: remainderIotaGoOutputID,
+			Output:   output,
+		}
+	} else {
+		// no funds remaining => no remainder output
+		f.lastRemainderOutput = nil
+	}
+	f.Unlock()
+
+	f.Events.IssuedMessage.Trigger(messageID)
+
+	return nil
+}
+
+// logSoftError logs a soft error and triggers the event.
+func (f *Faucet) logSoftError(err error) {
+	f.LogWarn(err)
+	f.Events.SoftError.Trigger(err)
+}
+
+// collectRequests collects faucet requests until the maximum amount or a timeout is reached.
+// locking not required.
+func (f *Faucet) collectRequests(ctx context.Context) ([]*queueItem, error) {
+
+	batchedRequests := []*queueItem{}
+	collectedRequestsCounter := 0
+
+CollectValues:
+	for collectedRequestsCounter < f.opts.maxOutputCount {
+		select {
+		case <-ctx.Done():
+			// faucet was stopped
+			return nil, common.ErrOperationAborted
+
+		case <-time.After(f.opts.batchTimeout):
+			// timeout was reached => stop collecting requests
+			break CollectValues
+
+		case <-f.flushQueue:
+			// flush signal => stop collecting requests
+			for collectedRequestsCounter < f.opts.maxOutputCount {
+				// collect all pending requests
+				select {
+				case request := <-f.queue:
+					batchedRequests = append(batchedRequests, request)
+					collectedRequestsCounter++
+
+				default:
+					// no pending requests
+					break CollectValues
+				}
+			}
+			break CollectValues
+
+		case request := <-f.queue:
+			batchedRequests = append(batchedRequests, request)
+			collectedRequestsCounter++
+		}
+	}
+
+	return batchedRequests, nil
+}
+
+// processRequestsWithoutLocking processes all possible requests considering the maximum transaction size and the remaining funds of the faucet.
+// write lock must be acquired outside.
+func (f *Faucet) processRequestsWithoutLocking(collectedRequestsCounter int, amount uint64, batchedRequests []*queueItem) []*queueItem {
+	processedBatchedRequests := []*queueItem{}
+	unprocessedBatchedRequests := []*queueItem{}
+	nodeAlmostSynced := f.nodeSyncedFunc() //TODO: check if we need an almost synced here or not
```

Comment:
I would say we don't need it.

There might be a small time window when a new milestone arrives in the node, where this check returns false.
But the requests will be readded to the queue and processed again after the batchTimeout anyway.

Maybe rename it to `nodeSynced` and remove the comment.